dex/src/utils.rs

```diff
@@ -0,0 +1,129 @@
+use std::iter::Peekable;
+
+use crate::{
+    critbit::{LeafNodeIterator, Slab},
+    matching::Side,
+};
+
+pub struct OrderBook<'a> {
+    side: Side,
+    slab: &'a Slab,
+}
+
+impl<'a> OrderBook<'a> {
+    pub fn new(side: Side, slab: &'a Slab) -> Self {
+        OrderBook { side, slab }
+    }
+
+    /// Iterate over level 2 data
+    pub fn levels(&'a self) -> LevelsIterator<'a> {
+        let is_descending = self.side == Side::Bid;
+        LevelsIterator {
+            leafs: self.slab.iter(is_descending).peekable(),
+        }
+    }
+}
```

Comment:
Think you can do this with a 1 liner if you import `itertools`:
```
self
    .slab
    .iter(is_descending)
    .map(|leaf| (leaf.price(), leaf.quantity())
    .group_by(|(p, q) *p)
    .map(|price, group| Level {price, quantity: group.map(|(_, q)| q).sum() })

```

Reply:
The only way i found that would satisfy lifetime requirements is to add an iter function on LevelsIterator, is this how we want it?

```rust
pub struct LevelsIterator<'a> {
    leafs: LeafNodeIterator<'a>,
}

impl<'a> LevelsIterator<'a> {
    fn iter(&'a self) -> Box<dyn Iterator<Item=Level> + 'a> {
        Box::new(self.leafs
            .map(|leaf| (leaf.price(), leaf.quantity()))
            .group_by(|(p, q)| *p)
            .into_iter()
            .map(|(price, group)| Level {
                price: price.get(),
                quantity: group.map(|(_, q)| q).sum(),
            }))
    }
}
```

Reply:
lgtm

Reply:
I ended up not adding this, i think while more elegant we end up bringing this heap allocation we didn't have before so any program using this will pay